### PR TITLE
fix: Teradata connection fixes

### PR DESCRIPTION
**Pull request by mccushjack**
### SUMMARY
fixes to Teradata driver and limit methods in teradata.py

110385813-a7590200-801c-11eb-95c4-ea227cd032c4
110385894-c6579400-801c-11eb-8a54-fe4e0f938157
110385924-cfe0fc00-801c-11eb-8b10-d2b169d3545e

### TEST PLAN
This solution works and is tested with Teradata Vantage, but we are happy to make other modifications if the team supports.

### ADDITIONAL INFORMATION
The request all realtes to the follwoing bug

New or Changed Public Interfaces
The present proposal requires no new or changed public interfaces.

### New dependencies
Only for Teradata connections, the present proposal requires the installation of

the teradatasqlalchemy connector, and
the teradatasql driver
packages to connect to a Teradata Vantage system Advanced SQL Engine Database.
These packages are actively maintained and their licenses are Teradata proprietary. The teradatasql license can be found at:
https://github.com/Teradata/python-driver/blob/master/LICENSE
whereas the teradatasqlalchemy code is not publicly posted at this point.

Migration Plan and Compatibility
No migration is necessary. The proposed code changes are an extension and non-destructive to the existing code, thus not affecting existing compatibility.

### CODE REVIEW

**Comment by villebro** [comment]
@mccushjack I'm going through old unreviewed PRs, and sadly this valuable PR had gone unreviewed. Can you rebase this PR to resolve the conflict? I'll finish reviewing this once that gets done.

**Comment by villebro** [comment]
@mccushjack oh no, this is conflicted again 🙁 Would you mind rebasing one more time and pinging me when you're done so I can get this merged?

**Comment by cwiebe18** [comment]
@mccushjack  thank you so much!

@junlincc @villebro anything you could do to nudge this along would be so appreciated!

**Comment by mccushjack** [comment]
@junlincc @villebro and @cwiebe18  -- The rebase is complete. The only major change outside of teradata.py fixes was an update to the main setup.py for loading Teradata's native python driver which removes the dependency on odbc driver. This should make interoperability easier and more seamless. LMK if you have any questions or comments.

**Comment by michael-s-molina** [comment]
@mccushjack I think something went wrong with the rebase. It's showing 1,099 file changes.

**Comment by mccushjack** [comment]
thanks, @michael-s-molina not sure what that was. made a few changes and checked https://github.com/apache/superset/pull/15930/files and it shows 5 files changed.

**Comment by betodealmeida** [comment]
This looks great!

Note that there are a few lint errors: https://github.com/apache/superset/runs/4072850679?check_suite_focus=true

You can fix some of these by running:

```
pip install pre-commit
pre-commit install
pre-commit run -a
```

**Review by betodealmeida — CHANGES_REQUESTED** [review]

File: superset/db_engine_specs/teradata.py
Comment (mccushjack) [reply superset/db_engine_specs/teradata.py:64]: Thanks, @betodealmeida should I create test_teradata in superset/tests/unit_tests/db_engine_specs/

Comment (betodealmeida) [reply superset/db_engine_specs/teradata.py:61]: Yeah, that would be perfect! Thanks! :)

Comment (mccushjack) [reply superset/db_engine_specs/teradata.py:61]: @betodealmeida @michael-s-molina  - I think I fixed the pylint errors and I added test_teradata.py to the unit tests. let me know if there are any other changes you need.

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:52]: Nit, these two `if` statements can be combined into a single one.

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:131]: Nit ,`child` might be clearer than `token2` here and in line 135.

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:196]:
```suggestion
        td_sel_keywork = {"SELECT ", "SEL "}
        td_limit_keywork = {"TOP", "SAMPLE"}
```

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:208]:
```suggestion
        tokens = str_statement.rstrip().split(" ")
```

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:209]:
```suggestion
        tokens = [token for token in tokens if token]
```

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:213]: Nit, traditionally `i` is used as an integer, `token` would be a better name.

File: tests/unit_tests/db_engine_specs/test_teradata.py
Comment (betodealmeida) [inline tests/unit_tests/db_engine_specs/test_teradata.py:32]:
```suggestion
    from superset.db_engine_specs.teradata.TeradataEngineSpec import apply_limit_to_sql
```

File: superset/db_engine_specs/teradata.py
Comment (betodealmeida) [inline superset/db_engine_specs/teradata.py:227]: Traditionally in Python we build strings by appending to a list (the performance for adding strings used to be terrible, theses days it's equivalent):

```suggestion
        result = []
        for token in new_tokens:
            result.append(token)
            if any(selword in token.upper() for selword in td_sel_keywork):
                token.append(f"TOP {final_limit}")
        return " ".join(result)
```

(untested)

**Review by betodealmeida — APPROVED** [review]
This looks great, thanks for the work!

I left a few comments on how to make the code more idiomatic, but I think this is also good to go as is. Your call, I know this has taken quite some time.

**Review by betodealmeida — CHANGES_REQUESTED** [review]
Actually, the unit test needs a fix before we can merge this.

File: superset/db_engine_specs/teradata.py
Comment (mccushjack) [reply superset/db_engine_specs/teradata.py]: @betodealmeida  -- I fixed the error in unit tests and made the other idiomatic changes you requested.

**Comment by villebro** [comment]
@mccushjack did you notice the failed unit test? I believe this is good to go once it's fixed:
image

**Comment by nytai** [comment]
@mccushjack seems like linting is failing for this PR. Sorry it has taken so long to get this change merged/reviewed. If you wouldn't mind fixing the lint errors and getting CI passing again, I'll make sure this gets merged. If you're part of the superset slack, feel free to ping me to get things expedited.

**Comment by mccushjack** [comment]
@nytai  I got some help to fix our unit test. I think everything is working now let @mccushjack  and @dmcnulla know if there are any updates needed.

**Comment by dmcnulla** [comment]
I do not know what this [Python Misc/pre-commit check](https://github.com/apache/superset/runs/5010947419?check_suite_focus=true) does, and I cannot tell from the error what is wrong. Can I buy a clue from somebody?

**Comment by nytai** [comment]
It runs a few linting type checks. @dmcnulla to get this check passing, the following steps should work.

- make an edits to `superset/tests/unit_tests/db_engine_specs/test_teradata.py` and `superset/superset/db_engine_specs/teradata.py` (can be simple blank space change)
- stage those files `git add . `
- run `pre-commit` (should be installed via `requirements/integration.txt`)

**Comment by nytai** [comment]
Closing, as this work was absorbed by https://github.com/apache/superset/pull/18240

**Pull request closed without merging**
